### PR TITLE
chore(sql): table_writer_metrics() and memory_metrics() for better observability

### PR DESCRIPTION
**Pull request by jerrinot**
It allows to query:
1. table writer metrics expose these: TOTAL_COMMITS, O3_COMMITS, ROLLBACKS, COMMITTED_ROWS, PHYSICALLY_WRITTEN_ROWS

2. memory utilization metrics: with a breakdown per tag. 

For #1 to work metrics have to be enabled in QuestDB configuration. #2 is always enabled. 

Questions I'd like reviewers to think about:
1.  This is effectively API. Are we OK to maintain compatibility? Is it ok to break it? Imagine we remove a memory tag.
2. Should we enable metrics by default? I feel stats on O3 could be pretty useful for troubleshooting
3. each table emits just a single row. more over table metrics are reset on QuestDB restart? is that OK? Alternatively, we could expose memory metrics as tuples where each row would represent a (key-value) pair. This would allow simple sorting by memory consumption.

### CODE REVIEW

**Comment by questdb-butler** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please update the PR title.
It should match this format: *type*: *description*
Where *type* is one of: "feat", "fix", "chore", "docs", "style", "refactor", "perf", "test", "build", "ci", "chore", "revert"
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5dc8c8e6cd9b379c3b32ebbe2aea243499cba99d
</p>

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 38 / 45 (84.44%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/TableWriterMetricsRecordCursorFactory.java|31|38|81.58%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/TableWriterMetricsFunctionFactory.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 51 / 53 (96.23%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/TableWriterMetricsRecordCursorFactory.java|44|46|95.65%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/TableWriterMetricsFunctionFactory.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 76 / 81 (93.83%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/SingleRowRecordCursor.java|19|22|86.36%|
|:large_blue_circle:|io/questdb/griffin/engine/table/MemoryMetricsRecordCursorFactory.java|19|20|95.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/TableWriterMetricsRecordCursorFactory.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/MemoryMetricsFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/TableWriterMetricsFunctionFactory.java|3|3|100.00%|

**Comment by jerrinot** [comment]
Please don't merge it yet, there is allocation going on due to auto-boxing. thanks to Vlad for spotting it!

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 74 / 77 (96.10%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/SingleLongRowRecordCursor.java|16|17|94.12%|
|:large_blue_circle:|io/questdb/griffin/engine/table/MemoryMetricsRecordCursorFactory.java|19|20|95.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/TableWriterMetricsRecordCursorFactory.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/MemoryMetricsFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/TableWriterMetricsFunctionFactory.java|3|3|100.00%|

**Comment by jerrinot** [comment]
@puzpuzpuz I agree with everything you wrote! I will change the format to return tuples and later send a separate PR to enable metrics by default.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 87 / 94 (92.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/StringLongTuplesRecordCursor.java|24|29|82.76%|
|:large_blue_circle:|io/questdb/griffin/engine/table/MemoryMetricsRecordCursorFactory.java|21|22|95.45%|
|:large_blue_circle:|io/questdb/griffin/engine/table/TableWriterMetricsRecordCursorFactory.java|31|32|96.88%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/MemoryMetricsFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/TableWriterMetricsFunctionFactory.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 91 / 94 (96.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/StringLongTuplesRecordCursor.java|26|29|89.66%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/TableWriterMetricsRecordCursorFactory.java|32|32|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/MemoryMetricsFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/MemoryMetricsRecordCursorFactory.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/TableWriterMetricsFunctionFactory.java|3|3|100.00%|

**Comment by jerrinot** [comment]
@puzpuzpuz can you please re-check?

File: core/src/test/java/io/questdb/griffin/engine/table/MemoryMetricsRecordCursorFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            }
+        });
+    }
+
+    private static void assertCursor(RecordCursor cursor) {
+        cursor.hasNext();
+        Record record = cursor.getRecord();
+
+        assertEquals("TOTAL_USED", record.getStr(0));
+        assertEquals(Unsafe.getMemUsed(), record.getLong(1));
```
[the remark this answers is not in the document]
Comment (jerrinot) [reply core/src/test/java/io/questdb/griffin/engine/table/MemoryMetricsRecordCursorFactoryTest.java:74]: true. but the exact comparison is inevitable if I want to use the shared testing infrastructure (`assertCursor()` and similar). the shared infra gives me confidence the factory/cursor does not misbehave. should I give up on that?

File: core/src/main/java/io/questdb/griffin/engine/functions/table/MemoryMetricsFunctionFactory.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.CursorFunction;
+import io.questdb.griffin.engine.table.MemoryMetricsRecordCursorFactory;
+import io.questdb.std.IntList;
+import io.questdb.std.ObjList;
+
+public final class MemoryMetricsFunctionFactory implements FunctionFactory {
+    @Override
+    public String getSignature() {
+        return "sys.memory_metrics()";
```
[the remark this answers is not in the document]
Comment (jerrinot) [reply core/src/main/java/io/questdb/griffin/engine/functions/table/MemoryMetricsFunctionFactory.java:40]: as discussed elsewhere: FunctionFactories do not have access to `CairoConfiguration`. JDK ServiceLoader instantiates factories by using a non-arg constructor. Andrey and I agreed on removing the prefix altogether.

File: core/src/test/java/io/questdb/griffin/engine/table/MemoryMetricsRecordCursorFactoryTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/test/java/io/questdb/griffin/engine/table/MemoryMetricsRecordCursorFactoryTest.java:74]: No problem with keeping the test as is. It's a nit comment after all.

**Review by puzpuzpuz — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 91 / 94 (96.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/StringLongTuplesRecordCursor.java|26|29|89.66%|
|:large_blue_circle:|io/questdb/cairo/TableWriterMetrics.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/TableWriterMetricsRecordCursorFactory.java|32|32|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/MemoryMetricsFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/MemoryMetricsRecordCursorFactory.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/TableWriterMetricsFunctionFactory.java|3|3|100.00%|

**Review by ideoma — COMMENTED** [review]
I'd prefer a function for all prometheus metrics, all in one, not just writer and memory separately.

**Pull request merged**